Document registering hosts to Foreman by running curl on the hosts (#359)
theforeman/foreman-documentation: guides/doc-Managing_Hosts/topics/proc_registering-a-host-to-project-using-the-global-registration-template.adoc

[id="registering-a-host-to-project-using-the-global-registration-template_{context}"]
= Registering a Host to {ProjectName} Using the Global Registration Template

Use the following procedure to register a host to {ProjectNameX}.

With this registration method you can register hosts to {Project} by generating a `curl` command on {Project} and running this command on hosts.
This method uses the global registration template that gives you complete control over the process of the host registration.

ifeval::["{build}" != "satellite"]
Note that you can extend the parameters by plug-ins.
For more information, see https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc[How to Create a Plugin].
endif::[]

.Prerequisites
* The {Project} user that generates the `curl` command must have the `create_hosts` permission.

> * You must have root privileges on the host that you want to register.

Reviewer: There is more if you plan to register RHEL hosts: activation key (subscription manager won't be happy without it), also if you plan to use automatic insights registration on that RHEL host, you'll need a rhel repo synced and made available through that activation key so that the host can install the insights-client from it

Author: Thank you, I have added those prerequisites

.Procedure
. In the {Project} web UI, navigate to *Hosts* > *Provisioning Templates*, find the *Linux registration default* template and click it.

. Click the *Association* tab.

> . Select the operating systems that you want to register.

Reviewer 2: this seems too brief; if the OS you want is already in the Selected items column then you will move it back to the All items column by selecting it. How about: 
Ensure the operating system you want is in the Selected items column. 
Click Submit to save your changes.

Reviewer: Not sure about other OSs but for RHEL the template is associated by default so this step wouldn't be needed unless customizing the template, worth double checking

Author: Added what @swadeley suggested. Ensuring is never redundant, and with this step we show users that the template must be set in the OS settings.

. Click *Submit*.

. Navigate to *Hosts* > *Operating Systems* and click the operating system that you want to register.

. Click the *Templates* tab.

> . From the *Registration template* list, select *Linux registration default*.

Reviewer 2: this is the default choice; are you required to describe it like this or can you say: 
From the *Registration template* list, ensure *Linux registration default* is selected.

Author: Done.

. Click *Submit*.

. Navigate to *Hosts* > *All Hosts* > *Register Host*.

. Optional: If you want to register hosts through {SmartProxy}, ensure that the *Registration* feature is enabled on this {SmartProxy}.
+
Navigate to *Infrastructure* > *{SmartProxies}*, click the {SmartProxy} that you want to use, and locate the *Registration* feature in the *Active features* list.
+
Optional: If the *Registration* feature is not enabled on your {SmartProxy}, enter the following command on the {SmartProxy} to enable it:
+
[options="nowrap", subs="+quotes,attributes"]
----
# foreman-installer --foreman-proxy-registration \
--foreman-proxy-templates \
--foreman-proxy-template-url 'http://{smartproxy-example-com}'
----

> . Optional: From the *Host Group* list, select the host group to associate the hosts with.

Reviewer 2: I am not sure this is optional, I know with bootstrap.py you had to include Host Group, need to test it

Reviewer 2: never mind, it works without HG, the activation key does the necessary

. From the *Operating System* list, select the operating system of hosts that you want to register.

. Optional: If you want to register hosts through {SmartProxy}, from the *{SmartProxy}* list, select the {SmartProxy} to register hosts to.

ifeval::["{build}" == "satellite"]
. From the *Insights* list, select whether you want to register the hosts to Insights or not.
endif::[]

. From the *Remote Execution* list, select whether you want to deploy {Project} SSH keys to hosts or not.

. Optional: {Project} uses the JSON Web Tokens (JWT) for authentication.

> This duration of this token defines how long the generated `curl` command works.

Reviewer 2: ```suggestion
The duration of this token defines how long the generated `curl` command works.
```

Author: Done.

Therefore, do not delete, block, or change permissions of the user during the token duration.

. Optional: In the *Remote Execution Interface* field, enter a network interface that hosts must use for the SSH connection.
If you keep this field blank, {Project} uses the default network interface.

ifeval::["{build}" == "satellite"]
. In the *Activation Key(s)* field, enter one or more activation keys to assign to registered hosts.
endif::[]

ifeval::["{build}" != "satellite"]
. Optional: This step is for the Katello users only.
If you register RHEL or CentOS hosts, in the *Activation Key(s)* field, enter one or more activation keys to assign to registered hosts.
endif::[]

. Click *Generate command*.

. Copy the generated `curl` command to enter it on the hosts.

. Configure hosts with the CA certificate with one of the following methods:
+
* Add the `--insecure` option to the `curl` command.
This makes the first call insecure, however, during this first call, hosts download the CA file from {Project}.
Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+

> If an attacker, locating in the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JWT.

Reviewer 2: ```suggestion
If an attacker, located on the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JWT.
```

Author: Done.

This method uses the global registration template that gives you complete control over the process of the host registration.

ifeval::["{build}" != "satellite"]
Note that you can extend the parameters by plug-ins. For more information, see https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc.
endif::[]

.Prerequisites
* The {Project} user that generates the `curl` command must have the `create_hosts` permission.
* You must have root privileges on the host that you want to register.

.Procedure
. In {Project} web UI, navigate to *Hosts* > *Provisioning Templates* and locate and click *Linux registration default*.

. Click the *Association* tab.

. Select the operating systems that you want to register.

. Navigate to *Hosts* > *Operating Systems* and click the operating system that you want to register.

. Click *Submit*.

. Click the *Templates* tab.

. From the *Registration template* list, select *Linux registration default*.

. Click *Submit*.

. Navigate to *Hosts* > *All Hosts* > *Register Host*.

> . Optional: If you want to register hosts to {SmartProxy}, enter the following command on the {SmartProxy} to enable required modules:

Author: did I get it right that users need to run this command on Capsule Server?

Reviewer 3: Yes, but I believe that registration module should be enabled by default.
So first step should be to check if capsule has enabled _Registration_ module (_Infrastructure > Capsules > refresh proxy_ & check if Registration feature is present), and if not, then enable it with the commands.

Author: Perfect, thank you

+
[options="nowrap", subs="+quotes,attributes"]
----
# foreman-installer --foreman-proxy-registration \
--foreman-proxy-templates \
--foreman-proxy-template-url 'http://{smartproxy-example-com}'
----

> . What fields are required for Satellite and Foreman?

Author: @stejskalleos I need your help with this question. We can schedule a call after the first review rounds.

Reviewer 3: Satellite: `Activation keys`
Foreman: 
RHEL & centOS: `Activation keys`
Other OS: `Hostgroup` or `operating system`

. Optional: From the *Host Group* list, select the host group to associate the hosts with.

. From the *Operating System* list, select the operating system of hosts that you want to register.

. Optional: If you want to register hosts to a {SmartProxy}, from the *{SmartProxy}* list, select the {SmartProxy} to register hosts to.

ifeval::["{build}" == "satellite"]

> . Optional: If you do not want to register hosts with Insights, from the *Insights* list, select *No*. What are the defaults for those two?

Author: @stejskalleos what are the default values for the Insights and Remote Execution settings in Satellite and Foreman? In my Satellite I just saw blank fields, maybe putting something there for visibility would help.

Reviewer 3: Logic for these two parameters is little bit confusing (it will be replaced in 6.10 with better behavior), I'll try my best:

**Options in Form**
* `empty option` - This means that we take the (default) value from global parameters `host_registration_insights` and `host_registration_remote_execution`. See bellow _Getting value of global parameters_
* `Yes / No` - If the selected value in form is same as the value of global parameter, we **do not**  add `setup_rex|setup_rex=value` to generated command. Only when the value is different. See the table:

Selected option in Form | Global parameter | Argument added to command
--- | --- | --- 
nothing | `true` | nothing (take value from GP)
nothing | `false` | nothing (take value from GP)
`Yes` | `true` | nothing (they're same)
`No` | `false` | nothing (they're same)
`Yes` | `false` | `...&setup_rex\|insights=true`
`No` | `true` | `...&setup_rex\|insights=false`

**Getting value of global parameters**
By default both parameters are set to `true` on global level. However, users can override it on multiple levels:

- Global Level (`insights=true`, `rex=true`)
- Organization
- Location
- Host Group
- Operating System

The rule is that lower level overrides the upper levels, see the example:

- Global Level (`insights=true`, `rex=true`)
- Organization
- Location
- Host Group (`insights=false`, `rex=false`)

Values of params: `false`

Author: In the recording you said that this logic is going to be reworked by the 6.9 release. Did you describe the latest planned versions?

Author: I will not mark this step as optional and just say that users need to select either Yes or Not to avoid any confusion. It is much easier to make users choose one concrete variant than describe this complicated logic.

endif::[]

. Optional: If you do not want to deploy {Project} SSH keys to hosts, from the *Remote Execution* list, select *No*.

. Optional: {Project} uses the JSON Web Tokens (JWT) for authentication.
This duration of this token defines how long the generated `curl` command works.
If you want to change the duration of the token, enter the required duration to the *Token lifetime (hours)* field.
You can set the duration to any number of hours.
+
Note that {Project} uses the permissions of the user who generated the `curl` command for authorization of hosts.
If user loses or gains additional permissions, the permissions of the JWT change too.
Therefore, do not delete, block, or change permissions of the user during the token duration.

> . Optional: What is remote execution interface?

Author: @stejskalleos, what is remote exectuion interface? What values can users put there and why?

Reviewer 3: It's network interface of the registered host machine. Some users may want to specify one specific virtual interface for ssh connections. More info here: https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking/

Author: Is this an optional field? What if users keep this feild blank?

Reviewer 3: yes, no need to fill the field.

ifeval::["{build}" != "satellite"]

> . In the *Activation Key(s)* field, enter one or more activation keys to assign to registered hosts.

Reviewer 2: ```suggestion
. In the *Activation Key(s)* field, enter one or more activation keys to assign to hosts.
```
because they are not actually registered at this point

Author: You are right, thank you

endif::[]

. Click *Generate command*.

. Copy the generated `curl` command to enter it on the hosts.

. On the hosts that you want to register, enter the generated `curl` command as `root`.

.SSL
For calling Foreman endpoints via https CA authority must be installed on the host.
Path to CA file can be found in Settings > Authentication > SSL CA file.

If you don't want to upload & install CA on the host, you can run first `curl` command

> with `---insecure` parameter, like this:

Author: @stejskalleos is it safe to tell Satellite users to run `curl` with the `insecure` option? Although we tell users to do this in the official Satellite docs [1] now, is there a better and safer way of doing this?

[1] See step 2 in https://access.redhat.com/documentation/en-us/red_hat_satellite/6.8/html-single/managing_hosts/index#registering-a-host-to-satellite

Reviewer 3: Well, there is always a risk that something could happen, but I would say it's okay.

If they don't like the idea of insecure call, they can manually install certification authority (File defined in _Settings > Authentication > SSL CA file_) on the registered machine and then they can securely call Satellite instance without `--insecure`

Author: How to install this CA file on hosts? Is that possible to describe this shortly or it is better to leave this decision on "how" to each user?

Reviewer 3: It's pretty easy:

* Download file from _Settings > Authentication > SSL CA file_
* Copy the file to the `/etc/pki/ca-trust/source/anchors/`
* Run command `update-ca-trust extract`

[options="nowrap", subs="+quotes,attributes"]
----
curl -X GET --insecure https://foreman.example.com/register | bash
----

> This make first call insecure, however Global Registration template will

Author: Sound like an attacker might fetch the CA from this call and mess something up. Is it that scary?

Reviewer 3: Well, to be honest, I don't know. But if attacker can fetch CA from the call, that means two things:

* He / She has access to the registered machine and from logs can get the command (with auth token)
* Or has access credentials to the API and can call /register endpoint

That being sad, having access to CA is the last thing we should worry about, IMHO.

Author: IDK here, an atacker accessing a registered host is not something I would like to see :) But I see your point, I will document both methods and details about the unsafety.
